enarx-keep-sev/src/vm/builder.rs

```diff
@@ -101,7 +217,67 @@ impl Builder {
         Ok(())
     }
 
-    pub fn build(self) -> VirtualMachine {
-        self.vm
+    fn relocate(component: &mut Component, offset: usize) {
+        component.entry += offset;
+
+        for seg in &mut component.segments {
+            seg.dst += offset;
+        }
+    }
+}
+
+/// Once both the shim and code have been loaded into the guest,
+/// we can finally configure the dual-purpose bootinfo page and
+/// prepare the startup vCPU.
+impl Builder<Code> {
+    pub fn build(self) -> Result<VirtualMachine, io::Error> {
+        let addr_space = self.data.address_space.as_ref().unwrap().as_virt();
+        let boot_info = self.data.boot_info.unwrap();
+
+        // To finish guest setup, certain values must be written into the guest
+        // address space via host-side pointers since the memory we are writing
+        // to is backed by the host address space.
+        let host_setup = unsafe {
+            let dst = addr_space.start;
+            let vm_setup = dst.as_mut_ptr::<MaybeUninit<VMSetup>>();
+            // zero the struct
+            vm_setup.write(MaybeUninit::<VMSetup>::zeroed());
+            &mut *dst.as_mut_ptr::<VMSetup>()
+        };
+        host_setup.shared_page = Page::copy(boot_info);
```

Comment:
`Page::copy()` copies an entire page, where the interesting part is only the `BootInfo` struct. The rest is zero anyway. This is why I added the `fill_page()` function. Anyway, it's only a copy of 4000 useless bytes and the CPU is fast enough.